### ISSUE TITLE
Git LFS files show as changed in mounted volume in Docker

### ISSUE THREAD

**Issue by KaloyanP** [git-lfs/git-lfs#3958]
**Describe the issue**
I do my development on OSX but I run all CI/CD tools inside a Linux Docker image. For local development, I am running the CI Docker image locally with my repository mounted at `/git`, so that I can run various commands in the exact environment that the CI process will run them in.

When I run `git status` on OSX, I get completely clean output. However, when I run the same command inside Docker, every single git lfs file gets picked up as changed.

I am trying to use git status to detect changes and do incremental builds, but this issue means that I can never get a clean status inside the Docker image.

I looked into differences in `.gitconfig`, ran `git lfs install` in both places, but I couldn't see a problem. Also, since the volume is mounted inside Docker, file contents should be *identical*.

I would welcome any help!

**System environment**
OSX

**Output of `git lfs env`**

Docker:
```
git-lfs/2.8.0 (GitHub; linux amd64; go 1.12.6)
git version 2.7.4
LocalWorkingDir=/git
LocalGitDir=/git/.git
LocalGitStorageDir=/git/.git
LocalMediaDir=/git/.git/lfs/objects
LocalReferenceDirs=
TempDir=/git/.git/lfs/tmp
ConcurrentTransfers=3
TusTransfers=false
BasicTransfersOnly=false
SkipDownloadErrors=false
FetchRecentAlways=false
FetchRecentRefsDays=7
FetchRecentCommitsDays=0
FetchRecentRefsIncludeRemotes=true
PruneOffsetDays=3
PruneVerifyRemoteAlways=false
PruneRemoteName=origin
LfsStorageDir=/git/.git/lfs
AccessDownload=none
AccessUpload=none
DownloadTransfers=basic
UploadTransfers=basic
git config filter.lfs.process = "git-lfs filter-process"
git config filter.lfs.smudge = "git-lfs smudge -- %f"
git config filter.lfs.clean = "git-lfs clean -- %f"
```

OSX:
```
git-lfs/2.7.2 (GitHub; darwin amd64; go 1.12.4)
git version 2.24.0

LocalWorkingDir=/git/repository
LocalGitDir=/git/repository/.git
LocalGitStorageDir=/git/repository/.git
LocalMediaDir=/git/repository/.git/lfs/objects
LocalReferenceDirs=
TempDir=/git/repository/.git/lfs/tmp
ConcurrentTransfers=3
TusTransfers=false
BasicTransfersOnly=false
SkipDownloadErrors=false
FetchRecentAlways=false
FetchRecentRefsDays=7
FetchRecentCommitsDays=0
FetchRecentRefsIncludeRemotes=true
PruneOffsetDays=3
PruneVerifyRemoteAlways=false
PruneRemoteName=origin
LfsStorageDir=/git/repository/.git/lfs
AccessDownload=none
AccessUpload=none
DownloadTransfers=basic
UploadTransfers=basic
GIT_EXEC_PATH=/usr/local/Cellar/git/2.24.0/libexec/git-core
git config filter.lfs.process = "git-lfs filter-process"
git config filter.lfs.smudge = "git-lfs smudge -- %f"
git config filter.lfs.clean = "git-lfs clean -- %f"
```

**Comment by bk2204** [comment]
Hey,

Sorry to hear you're having trouble.  Inside the container, when you see the modified files, can you run `git diff --ws-error-highlight=all` and provide an example of one of the diffs?  Can you also provide your `.gitattributes` file and the output of `git config -l` both inside and outside the container?

When you're using a container, the stat information inside the container differs from the information outside, so Git will re-read all of the files since the stat information in the index is out of date.  (You may be able to avoid this by setting `core.checkStat` to `minimal`.)  However, there's something that differs in the environment inside and outside the container that's causing Git to think the data is different when it re-reads and re-filters the files, and the information I've asked for will help track down what that is.

**Comment by KaloyanP** [comment]
Thank you so much for offering to help!

I rand the commands and also compared `git config -l` output (I have included the full output after this), and the only difference was a few extra config lines on OSX that are used by tooling, etc., as well as the `core.autocrlf` option. I also tried to make diffs appear on OSX with all options of `core.autocrlf`, but files would always get detected as differing between the two. I also ran everything with the latest versions of `git` and `git-lfs` installed inside the Docker image, but output was exactly the same.

`git diff --ws-error-highlight=all ./image.png` *:*

OSX (did not output anything):
```
$ git diff --ws-error-highlight=all ./image.png
```

Docker (not very exciting output): 
```
$ git diff --ws-error-highlight=all ./image.png
diff --git a/image.png b/image.png
index 7ab4514..aff8cef 100644
Binary files a/image.png and b/image.png differ
```
**.gitattributes**
```
image.png filter=lfs diff=lfs merge=lfs -text
*.avi filter=lfs diff=lfs merge=lfs -text
*.caffemodel filter=lfs diff=lfs merge=lfs -text
*.mp4 filter=lfs diff=lfs merge=lfs -text
*.jpg filter=lfs diff=lfs merge=lfs -text
*.m4a filter=lfs diff=lfs merge=lfs -text
*.wav filter=lfs diff=lfs merge=lfs -text
*.png filter=lfs diff=lfs merge=lfs -text
*.svg filter=lfs diff=lfs merge=lfs -text
*.woff filter=lfs diff=lfs merge=lfs -text
*.otf filter=lfs diff=lfs merge=lfs -text
**.h5 filter=lfs diff=lfs merge=lfs -text
```

**`git config -l (extra lines on OSX)**
```credential.helper=osxkeychain
user.name=KaloyanP
user.email=KaloyanP@example.com
core.excludesfile=/Users/KaloyanP/.gitignore_global
core.autocrlf=false
core.editor=code -n -w
difftool.sourcetree.cmd=opendiff "$LOCAL" "$REMOTE"
difftool.sourcetree.path=
mergetool.sourcetree.cmd=...
mergetool.sourcetree.trustexitcode=true
commit.template=/Users/KaloyanP/.stCommitMsg```

**`git config -l (OSX)`**
```
credential.helper=osxkeychain
user.name=KaloyanP
user.email=KaloyanP@example.com
core.excludesfile=/Users/KaloyanP/.gitignore_global
core.autocrlf=false
core.editor=code -n -w
difftool.sourcetree.cmd=opendiff "$LOCAL" "$REMOTE"
difftool.sourcetree.path=
mergetool.sourcetree.cmd=...
mergetool.sourcetree.trustexitcode=true
commit.template=/Users/KaloyanP/.stCommitMsg
filter.lfs.clean=git-lfs clean -- %f
filter.lfs.smudge=git-lfs smudge -- %f
filter.lfs.process=git-lfs filter-process
filter.lfs.required=true
core.repositoryformatversion=0
core.filemode=true
core.bare=false
core.logallrefupdates=true
core.ignorecase=true
core.precomposeunicode=true
remote.origin.url=git@bitbucket.org:...
remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
branch.master.remote=origin
branch.master.merge=refs/heads/master
lfs.https://bitbucket.org/...
```

**`git config -l (Docker)`** 
```
filter.lfs.clean=git-lfs clean -- %f
filter.lfs.smudge=git-lfs smudge -- %f
filter.lfs.process=git-lfs filter-process
filter.lfs.required=true
core.autocrlf=input
core.repositoryformatversion=0
core.filemode=true
core.bare=false
core.logallrefupdates=true
core.ignorecase=true
core.precomposeunicode=true
remote.origin.url=git@bitbucket.org:...
remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
branch.master.remote=origin
branch.master.merge=refs/heads/master
lfs.https://bitbucket.org/...
```

**Comment by bk2204** [comment]
> Docker (not very exciting output):
> 
> ```
> $ git diff --ws-error-highlight=all ./image.png
> diff --git a/image.png b/image.png
> index 7ab4514..aff8cef 100644
> Binary files a/image.png and b/image.png differ
> ```

This is actually very interesting.  On my system, these produce a diff like this:

```
diff --git a/temp/ghi.bin b/temp/ghi.bin
index e9eecd0..0b954e4 100644
--- a/temp/ghi.bin
+++ b/temp/ghi.bin
@@ -1,3 +1,3 @@
 version https://git-lfs.github.com/spec/v1
-oid sha256:8807953138274f7bd417673cde9848823b1fe720b5ed7b01f8fa00681a456da0
+oid sha256:76bd153c0f9608abb5bd4e4e96cc571fe064f75fb0003e7574cf3bc369da0b3e
 size 4
```

So I'd expect to see something like that here as well, even though the files are marked as `-text`.  Does doing a `git add --renormalize .` on OS X fix cause anything to be changed?  If so, does committing that and then entering the container work for you as expected?

If not, does adding the `-a ` flag to the `git diff` invocation you used above cause anything like my output to be printed?

**Comment by KaloyanP** [comment]
Apologies for the radio silence, I hope you have had a great time over the holiday season!

I decided to clone the repository again, and what do you know - a whole bunch of warnings about "Files should have been pointers but weren't". Then I ran `git add --renormalize .` and it wasn't a clean output - so I was on to something.

The solution was to add these files (2000 of them, no less 🤦‍♂) to LFS. This resolved the problem.

I think this would have been avoided if one of the steps I ran above had output "Found files that should have been pointers but weren't" in my main working repository without having to clone it or run Docker on it. However, I have provided the full sequence of events above in case you need it to inform your product development :-).

It turned out the problem were some files that "should have been pointers but weren't". The best way I understand the sequence of events is that:
1. We started using git-lfs for some big, multi-GB files without migrating the rest of the repository
2. We continued adding more and more binaries to LFS
3. At some point, someone decided to apply blanket filters, such as the ones above
4. However, at the same time, someone decided to commit multiple binaries to `git` that would have been covered by the filters above - but they didn't know how to use `LFS`
5. This immediately caused some issues, prompting "Found some files that should have been pointers but weren't"
6. The warning messages were ignored at the time because, well, we were new to LFS and warnings were to be expected. Those warning messages on OSX were automatically hidden through some git mechanism that I don't fully understand. However, they would be present every time the repository was cloned. *This is the state the Docker image seems to be in every time - as it if has just cloned the repository*
8. We never use git within the Docker CI image because it's painfully slow on OSX, so we never noticed the issue in 7, until we decided to use git hashes for incremental builds.

TLDR; The solution was to add the files-that-should-have-been-pointers-but-weren't (2000 of them, no less 🤦‍♂) to LFS. This resolved the problem.

**Comment by bk2204** [comment]
Hey,

I'm glad you figured this out.  This is a tricky edge case, and it bites a lot of people.

Unfortunately, there's a limit to how much we can do in this particular case, since we're at the mercy of Git's filtering mechanisms in Git LFS, and therefore we can't do much more than warn.  Aborting is the alternative, and in this case would make it harder to fix, so we don't really want to do that.

Users can always skip the filtering, as you've seen, although we have an item on our project board for future Git features that makes filters mandatory, so that users will see an error if Git LFS is installed.

Since we know what needs to be done here, I'm going to close this, but if you have any more problems, please feel free to reply back or open a new issue.
